compiler/kernel/bufferfilter.go

```diff
@@ -73,41 +80,47 @@ func CompileBufferFilter(e dag.Expr) (*expr.BufferFilter, error) {
 	}
 }
 
-func isFieldEqualOrIn(e *dag.BinaryExpr) (*astzed.Primitive, string) {
+// XXX isFieldEqualOrIn should work for any paths not just top-level fields.
+// See issue #3412
+
+func isFieldEqualOrIn(zctx *zed.Context, e *dag.BinaryExpr) (*zed.Value, string) {
 	if dag.IsTopLevelField(e.LHS) && e.Op == "=" {
-		if literal, ok := e.RHS.(*astzed.Primitive); ok {
-			return literal, "="
+		if literal, ok := e.RHS.(*dag.Literal); ok {
+			val, err := zson.ParseValue(zctx, literal.Value)
+			if err != nil {
+				return nil, ""
+			}
+			return &val, "="
 		}
 	} else if dag.IsTopLevelField(e.RHS) && e.Op == "in" {
-		if literal, ok := e.LHS.(*astzed.Primitive); ok && literal.Type != "net" {
-			return literal, "in"
+		if literal, ok := e.LHS.(*dag.Literal); ok {
+			val, err := zson.ParseValue(zctx, literal.Value)
+			if err != nil {
+				return nil, ""
+			}
+			if val.Type == zed.TypeNet {
+				return nil, ""
+			}
+			return &val, "="
 		}
 	}
 	return nil, ""
 }
 
-func newBufferFilterForLiteral(l astzed.Primitive) (*expr.BufferFilter, error) {
-	switch l.Type {
-	case "bool", "byte", "int16", "uint16", "int32", "uint32", "int64", "uint64",
-		"float32", "float64", "time", "duration":
+func newBufferFilterForLiteral(val *zed.Value) (*expr.BufferFilter, error) {
+	switch val.Type.(type) {
+	case *zed.TypeOfBool, *zed.TypeOfInt8, *zed.TypeOfInt16, *zed.TypeOfUint16, *zed.TypeOfInt32, *zed.TypeOfUint32, *zed.TypeOfInt64, *zed.TypeOfUint64, *zed.TypeOfFloat32, *zed.TypeOfFloat64, *zed.TypeOfTime, *zed.TypeOfDuration:
 		// These are all comparable, so they can require up to three
 		// patterns: float, varint, and uvarint.
 		return nil, nil
-	case "null":
-		return nil, nil
-	case "regexp":
-		// Could try to extract a pattern (e.g., "efg" from "(ab|cd)(efg)+[hi]").
+	case *zed.TypeOfNull:
 		return nil, nil
-	case "string":
+	case *zed.TypeOfString:
 		// Match the behavior of zed.ParseLiteral.
-		l.Type = "bstring"
-	}
-	v, err := zson.ParsePrimitive(l.Type, l.Text)
-	if err != nil {
-		return nil, err
+		val = zed.NewValue(zed.TypeBstring, val.Bytes)
 	}
```

Comment:
1. The first case is missing `zed.TypeOfUint8` and shouldn't include `zed.TypeOfBool` since it doesn't require multiple patterns. We should just use `zed.IsNumber`.
2. There's no need to turn Zed strings into bstrings since we're no longer trying to get `zson.ParsePrimitive` to behave like `zed.ParseLiteral`.
3. Might as well armor this against named types.
```go
	if id := val.Type.ID(); zed.IsNumber(id) || id == zed.IDNull {
		// All numbers are all comparable, so they can require up to three
		// patterns: float, varint, and uvarint.
		return nil, nil
	}
```